content/5_Plat/2DRQ.mdx

```diff
@@ -259,58 +276,67 @@ easier to understand, albeit significantly slower due to a high constant factor:
 <CPPSection>
 
 ```cpp
-// index of largest value <= x in v (sorted)
-// if v = [1, 2, 4], ind(v, 3) would return 1
-int ind(vector<int> v, int x) {
-	return upper_bound(v.begin(), v.end(), x) - v.begin() - 1;
-}
+#include <bits/stdc++.h>
+using namespace std;
+using ll = long long;
 
-class BIT2D {
+/**
+ * Offline 2D Fenwick Tree implementation.
+ * Note that n needs to be of a reasonable size, and
```

Comment:
what does reasonable size mean

Reply:
maybe i should just specify that the rows aren't coordinate compressed in the template itself

Reply:
perhaps smth smth "memory takes up is O(n^2) actually haha"

Reply:
no it doesnt take O(n^2) memory tho lmao, which is why its actually kinda useful

its like O(n log n) memory or similar, depending on the # of update queries

Reply:
well i guess n=1e5 would work?

Reply:
yeah n=1e5 won’t tle

Reply:
i mean then i'm not sure why this warning is necessary
when does this start breaking down

Reply:
i'd guess n=1e6 or n=1e7

the main reason is because the # of columns can scale to a stupid big number, but we aren't coordinate compressing the rows here for the sake of making the implementation easier